### PR TITLE
Implemented VolatileLayerClient::GetData() method

### PR DESCRIPTION
**Pull request by meravingen123**
Added Tests for VolatileLayerClientImpl class

Resolves: OLPEDGE-760

Signed-off-by: Serhii Lozynskyi <ext-serhii.lozynskyi@here.com>

### CODE REVIEW

File: olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp [hunk 1]
```diff
@@ -19,20 +19,89 @@
...
+}
+
 VolatileLayerClientImpl::VolatileLayerClientImpl(
     olp::client::HRN catalog, std::string layer_id,
-    olp::client::OlpClientSettings client_settings) {}
+    olp::client::OlpClientSettings client_settings)
+    : catalog_(std::move(catalog)),
+      layer_id_(std::move(layer_id)),
+      client_settings_(std::move(client_settings)),
+      pending_requests_(new PendingRequests()) {}
```
Comment (mykhailo-kuchma) [inline olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp]: I would suggest using make_shared

File: olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp [hunk 2]
```diff
@@ -19,20 +19,89 @@
...
+    }
+
+    auto blob_response = repository::DataRepository::GetBlobData(
+        catalog_, layer_id_, "volatile-blob", data_request, context,
+        client_settings_);
+
+    pending_requests->Remove(request_key);
+    if (callback) {
+      callback(blob_response);
+    }
```
Comment (mykhailo-kuchma) [inline olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp]: We can override a call to callback and call pending_requests->Remove() one time.
Fetch option Cache with update is not handled.

File: olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp [hunk 3]
```diff
@@ -19,19 +19,107 @@
...
 olp::client::CancellationToken VolatileLayerClientImpl::GetData(
     DataRequest data_request, Callback callback) {
-  return olp::client::CancellationToken{[]() {}};
+  auto catalog = catalog_;
+  auto client_settings = client_settings_;
+  auto layer_id = layer_id_;
+
+  auto get_data = [=](DataRequest data_request,
+                      olp::client::CancellationContext context,
+                      Callback callback) mutable {
```
Comment (borzun) [inline olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp]:
```suggestion
                      Callback callback) {
```
Should it be mutable?

File: olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp [hunk 4]
```diff
@@ -19,19 +19,107 @@
...
+                               ? data_request.GetPartitionId().get().c_str()
+                               : "<none>");
+        if (callback) {
+          callback(client::ApiError(client::ErrorCode::NotFound,
+                                    "Partition not found"));
+        }
+        return;
+      }
+
+      data_request.WithDataHandle(partitions.front().GetDataHandle());
```
Comment (borzun) [inline olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp]: Just curious - can we have more than 1 partition with same id?

File: olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp [hunk 5]
```diff
@@ -19,19 +19,107 @@
...
+    }
+
+    auto blob_response = repository::DataRepository::GetBlobData(
+        catalog_, layer_id_, "volatile-blob", data_request, context,
```
Comment (borzun) [inline olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp]:
```suggestion
        catalog, layer_id, "volatile-blob", data_request, context,
```

File: olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp [hunk 6]
```diff
@@ -19,19 +19,107 @@
...
+        client_settings_);
```
Comment (borzun) [inline olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp]:
```suggestion
        client_settings);
```

File: olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp (hunk 4 above)
Comment (meravingen123) [reply olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp]: This should not happen as i understand.
@mykhailo-kuchma Can you please clarify?

File: olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp [hunk 7]
```diff
@@ -19,19 +19,107 @@
...
+}
 
 olp::client::CancellationToken VolatileLayerClientImpl::GetData(
     DataRequest data_request, Callback callback) {
-  return olp::client::CancellationToken{[]() {}};
+  auto catalog = catalog_;
+  auto client_settings = client_settings_;
+  auto layer_id = layer_id_;
+
+  auto get_data = [=](DataRequest data_request,
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp]:
```suggestion
  auto get_data = [=](DataRequest request,
```

File: olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp [hunk 8]
```diff
@@ -19,19 +19,107 @@
...
+  auto add_task = [&](DataRequest& data_request, Callback callback) mutable {
+    olp::client::CancellationContext context;
+    auto cancellation_token = olp::client::CancellationToken(
+        [context]() mutable { context.CancelOperation(); });
+
+    int64_t request_key = pending_requests_->GenerateRequestPlaceholder();
+    pending_requests->Insert(cancellation_token, request_key);
+
+    repository::ExecuteOrSchedule(&client_settings, [=]() mutable {
+      get_data(data_request, context, std::move(callback));
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp]: A little bit to many copies of the callback.
Why do we need it? Can we not just capture it?

File: olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp [hunk 9]
```diff
@@ -19,19 +19,107 @@
...
+        catalog, layer_id, "volatile-blob", data_request, context,
+        client_settings);
+
+    if (callback) {
+      callback(blob_response);
+    }
+  };
+
+  auto pending_requests = pending_requests_;
+  auto add_task = [&](DataRequest& data_request, Callback callback) mutable {
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp]: why mutable?
You only alter pending_requests which is a shared_ptr so you should be able to remove the mutable.

File: olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp [hunk 10]
```diff
@@ -19,19 +19,107 @@
...
+  };
+
+  if (data_request.GetFetchOption() == FetchOptions::CacheWithUpdate) {
+    auto cache_token =
+        add_task(data_request.WithFetchOption(FetchOptions::CacheOnly),
+                 std::move(callback));
+    auto online_token = add_task(
+        data_request.WithFetchOption(FetchOptions::OnlineOnly), nullptr);
+
+    return olp::client::CancellationToken([=]() {
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp]: We don't need to add this in the pending_requests also?

File: olp-cpp-sdk-dataservice-read/tests/VolatileLayerClientImplTest.cpp [hunk 11]
```diff
@@ -18,24 +18,168 @@
...
+      olp::client::ApiResponse<VolatileLayerClientImpl::DataResult,
+                               olp::client::ApiError>;
+  std::promise<DataResponse> promise;
+  std::future<DataResponse> future = promise.get_future();
+
+  auto token = client.GetData(
+      request, [&](DataResponse response) { promise.set_value(response); });
+
+  const auto& response = future.get();
+  ASSERT_FALSE(response.IsSuccessful());
```
Comment (mykhailo-kuchma) [inline olp-cpp-sdk-dataservice-read/tests/VolatileLayerClientImplTest.cpp]: Please add assertion on the response code, as the user may rely on it

File: olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp (hunk 8 above)
Comment (mykhailo-kuchma) [reply olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp]: There is no easy way to move capture it in lambda.

Comment (andescu) [reply olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp]: As you already copy capture the callback you actually only move the copy.
At least in most of the cases. So if you already copied it you can work with the copy directly.

**Pull request merged**